sdk/monitor/Azure.Monitor.Ingestion/src/Models/UploadLogsOptions.cs

```diff
@@ -21,5 +28,62 @@ public class UploadLogsOptions
         /// <remarks> In the Upload method, this parameter is not used as the batches are uploaded in sequence. For parallel uploads, if this value is not set the default concurrency will be 5. </remarks>
         /// </summary>
         public int MaxConcurrency { get; set; }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        public event SyncAsyncEventHandler<UploadFailedEventArgs> UploadFailedEventHandler;
+
+        internal ClientDiagnostics _clientDiagnostics;
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="uploadFailedArgs"></param>
+        /// <param name="clientDiagnostics"></param>
+        /// <returns></returns>
+        internal virtual async Task InvokeEvent(UploadFailedEventArgs uploadFailedArgs, ClientDiagnostics clientDiagnostics)
+        {
+            await UploadFailedEventHandler.RaiseAsync(uploadFailedArgs, nameof(LogsIngestionClient), "Upload", clientDiagnostics).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="async"></param>
+        /// <param name="eventArgs"></param>
+        /// <param name="options"></param>
+        internal virtual async Task OnExceptionAsync(bool async, UploadFailedEventArgs eventArgs, UploadLogsOptions options)
```

Comment:
I'm not sure if I can make OnExceptionAsync and InvokeEvent protected virtual methods because InvokeEvent takes in the ClientDiagnostics which is an internal type. I can make the OnExceptionAsync method protected and keep InvokeEvent as internal?

Reply:
InvokeEvent should not be protected. In fact, you can probably just delete that method. In terms of getting the ClientDiagnostics into the options, we can use an internal property to set it.

Reply:
Alternatively, I would be fine with punting this protected method for now for the sake of time.

Reply:
Alternatively, you can add an internal `ClientDiagnostics` to the `EventArgs` if you want to the `Foo` event to have an `OnFoo` virtual.

Reply:
Yeah that is probably cleaner.